pkg/scanner/scanner.go

```diff
@@ -239,41 +294,99 @@ func (s *Scanner) Results() *types.ScanResults {
 	}
 }
 
-func (s *Scanner) HandleResult(result *forwarding.ImageVulnerabilities) error {
+func (s *Scanner) shouldIgnoreResult(scanD *scanData, resultScanUUID string, image string) bool {
+	if scanD == nil {
+		log.WithFields(s.logFields).Warnf("no scan data for image '%v', probably an old scan result - ignoring", image)
+		return true
+	}
+
+	if resultScanUUID != scanD.scanUUID {
+		log.WithFields(s.logFields).Warnf("Scan UUID mismatch, probably an old scan result - ignoring. image=%v, received=%v, expected=%v", image, resultScanUUID, scanD.scanUUID)
+		return true
+	}
+
+	if scanD.timeout {
+		log.WithFields(s.logFields).Warnf("Scan result after timeout - ignoring. image=%v, scan uuid=%v", image, resultScanUUID)
+		return true
+	}
+
+	if scanD.completed {
+		log.WithFields(s.logFields).Warnf("Duplicate result for image scan. image=%v, scan uuid=%v", image, resultScanUUID)
+		return true
+	}
+
+	return false
+}
+
+func (s *Scanner) HandleVulnerabilitiesResult(result *forwarding.ImageVulnerabilities) error {
 	s.Lock()
 	defer s.Unlock()
 
 	scanD, ok := s.imageToScanData[result.Image]
-	if !ok || scanD == nil {
-		log.WithFields(s.logFields).Warnf("no scan data for image '%v', probably an old scan result - ignoring", result.Image)
+	if !ok || s.shouldIgnoreResult(scanD, result.ScanUUID, result.Image) {
+		log.WithFields(s.logFields).Warnf("Ignoring vulnerabilities result for image '%v'", result.Image)
 		return nil
 	}
 
-	if result.ScanUUID != scanD.scanUUID {
-		log.WithFields(s.logFields).Warnf("Scan UUID mismatch, probably an old scan result - ignoring. image=%v, received=%v, expected=%v", result.Image, result.ScanUUID, scanD.scanUUID)
-		return nil
+	vulnerabilitiesResult := &vulnerabilitiesScanResult{
+		result:     result.Vulnerabilities,
+		success:    result.Success,
+		completed:  true,
+		scanErrMsg: result.ScanErrMsg,
 	}
 
-	if scanD.timeout {
-		log.WithFields(s.logFields).Warnf("Scan result after timeout - ignoring. image=%v, scan uuid=%v", result.Image, result.ScanUUID)
+	scanD.setVulnerabilitiesResult(vulnerabilitiesResult)
+
+	if scanD.vulnerabilitiesResult.success && scanD.vulnerabilitiesResult.result == nil {
+		log.WithFields(s.logFields).Infof("No vulnerabilities found on image %v.", result.Image)
+	}
+	if !scanD.vulnerabilitiesResult.success {
+		log.WithFields(s.logFields).Warnf("Scan of image %v has failed: %v", result.Image, scanD.vulnerabilitiesResult.scanErrMsg)
```

Comment:
Maybe we should state in the log that `vulnerabilities scan of ...`

* same for dockerfile scan